# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/Document.cpp
# Commit: 7af1b30df51c
# Full Hash: 7af1b30df51cd14f0b0fe9c1b206c78515c67677
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2019-06-10 21:47:35
# Regressor Bug: 1529884
# File Overlap Count: 1
# Description:
#   Bug 1529884 - part 4: Make Document::ExecCommand() do security check first if given command is supported r=smaug
#   
#   This minimize the next patch.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D29631
# ==============================================================================

diff -r 460df951943a -r 7af1b30df51c dom/base/Document.cpp
--- a/dom/base/Document.cpp	Mon Jun 10 10:24:59 2019 +0000
+++ b/dom/base/Document.cpp	Mon Jun 10 10:25:31 2019 +0000
@@ -4286,8 +4286,7 @@
     return false;
   }
 
-  // special case for cut & copy
-  // cut & copy are allowed in non editable documents
+  // Do security check first.
   if (commandData.IsCutOrCopyCommand()) {
     if (!nsContentUtils::IsCutCopyAllowed(&aSubjectPrincipal)) {
       // We have rejected the event due to it not being performed in an
@@ -4298,7 +4297,16 @@
                                       "ExecCommandCutCopyDeniedNotInputDriven");
       return false;
     }
-
+  } else if (commandData.IsPasteCommand()) {
+    if (!nsContentUtils::PrincipalHasPermission(&aSubjectPrincipal,
+                                                nsGkAtoms::clipboardRead)) {
+      return false;
+    }
+  }
+
+  // special case for cut & copy
+  // cut & copy are allowed in non editable documents
+  if (commandData.IsCutOrCopyCommand()) {
     // For cut & copy commands, we need the behaviour from
     // nsWindowRoot::GetControllers which is to look at the focused element, and
     // defer to a focused textbox's controller The code past taken by other
@@ -4318,12 +4326,6 @@
     return false;
   }
 
-  if (commandData.IsPasteCommand() &&
-      !nsContentUtils::PrincipalHasPermission(&aSubjectPrincipal,
-                                              nsGkAtoms::clipboardRead)) {
-    return false;
-  }
-
   // get command manager and dispatch command to our window if it's acceptable
   RefPtr<nsCommandManager> commandManager = GetMidasCommandManager();
   if (!commandManager) {
